File: sdks/go/verifier/verifier_test.go
package verifier_test

import (
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	"github.com/richardadalton/settled/sdks/go/verifier"
)

func vectorsDir() string {
	_, file, _, _ := runtime.Caller(0)
	return filepath.Join(filepath.Dir(file), "../../../test-vectors")
}

func loadJSON(name string) []byte {
	data, err := os.ReadFile(filepath.Join(vectorsDir(), name))
	if err != nil {
		panic(fmt.Sprintf("cannot read %s: %v", name, err))
	}
	return data
}

func h(s string) []byte {
	b, err := hex.DecodeString(s)
	if err != nil {
		panic(err)
	}
	return b
}

func b32(s string) [32]byte {
	var out [32]byte
	copy(out[:], h(s))
	return out
}

func proofs(hexes []string) [][32]byte {
	out := make([][32]byte, len(hexes))
	for i, s := range hexes {
		out[i] = b32(s)
	}
	return out
}

// ── Leaf hashes ───────────────────────────────────────────────────────────────

func TestLeafHash(t *testing.T) {
	var vectors []struct {
		Description string `json:"description"`
		InputHex    string `json:"input_hex"`
		HashHex     string `json:"hash_hex"`
	}
	json.Unmarshal(loadJSON("leaf-hashes.json"), &vectors)

	for _, v := range vectors {
		t.Run(v.Description, func(t *testing.T) {
			got := verifier.LeafHash(h(v.InputHex))
			if hex.EncodeToString(got[:]) != v.HashHex {
				t.Fatalf("got %x, want %s", got, v.HashHex)
			}
		})
	}
}

// ── Node hashes ───────────────────────────────────────────────────────────────

func TestNodeHash(t *testing.T) {
	var vectors []struct {
		Description    string `json:"description"`
		LeftHex        string `json:"left_hex"`
		RightHex       string `json:"right_hex"`
		HashHex        string `json:"hash_hex"`
		SwappedHashHex string `json:"swapped_hash_hex"`
	}
	json.Unmarshal(loadJSON("node-hashes.json"), &vectors)

	for _, v := range vectors {
		v := v
		if v.HashHex != "" {
			t.Run(v.Description, func(t *testing.T) {
				got := verifier.NodeHash(b32(v.LeftHex), b32(v.RightHex))
				if hex.EncodeToString(got[:]) != v.HashHex {
					t.Fatalf("got %x, want %s", got, v.HashHex)
				}
			})
		}
		if v.SwappedHashHex != "" {
			t.Run(v.Description+" non-commutative", func(t *testing.T) {
				ab := verifier.NodeHash(b32(v.LeftHex), b32(v.RightHex))
				ba := verifier.NodeHash(b32(v.RightHex), b32(v.LeftHex))
				if ab == ba {
					t.Fatal("node_hash must not be commutative")
				}
				if hex.EncodeToString(ba[:]) != v.SwappedHashHex {
					t.Fatalf("swapped: got %x, want %s", ba, v.SwappedHashHex)
				}
			})
		}
	}
}

// ── Inclusion proofs ──────────────────────────────────────────────────────────

func TestVerifyInclusion(t *testing.T) {
	var vectors []struct {
		TreeSize    uint64   `json:"tree_size"`
		LeafIndex   uint64   `json:"leaf_index"`
		LeafHashHex string   `json:"leaf_hash_hex"`
		ProofHex    []string `json:"proof_hex"`
		RootHex     string   `json:"root_hex"`
	}
	json.Unmarshal(loadJSON("inclusion-proofs.json"), &vectors)

	for _, v := range vectors {
		v := v
		t.Run(fmt.Sprintf("size=%d idx=%d", v.TreeSize, v.LeafIndex), func(t *testing.T) {
			ok := verifier.VerifyInclusion(b32(v.LeafHashHex), v.LeafIndex, v.TreeSize, proofs(v.ProofHex), b32(v.RootHex))
			if !ok {
				t.Fatal("verify_inclusion returned false")
			}
		})
	}
}

// ── Consistency proofs ────────────────────────────────────────────────────────

func TestVerifyConsistency(t *testing.T) {
	var vectors []struct {
		OldSize    uint64   `json:"old_size"`
		NewSize    uint64   `json:"new_size"`
		OldRootHex string   `json:"old_root_hex"`
		NewRootHex string   `json:"new_root_hex"`
		ProofHex   []string `json:"proof_hex"`
	}
	json.Unmarshal(loadJSON("consistency-proofs.json"), &vectors)

	for _, v := range vectors {
		v := v
		t.Run(fmt.Sprintf("old=%d new=%d", v.OldSize, v.NewSize), func(t *testing.T) {
			ok := verifier.VerifyConsistency(v.OldSize, v.NewSize, proofs(v.ProofHex), b32(v.OldRootHex), b32(v.NewRootHex))
			if !ok {
				t.Fatal("verify_consistency returned false")
			}
		})
	}
}

// ── Signed Tree Heads ─────────────────────────────────────────────────────────

func TestVerifyTreeHead(t *testing.T) {
	var vectors []struct {
		Description  string `json:"description"`
		TreeSize     uint64 `json:"tree_size"`
		RootHashHex  string `json:"root_hash_hex"`
		TimestampNs  int64  `json:"timestamp_ns"`
		SignatureHex string `json:"signature_hex"`
		PublicKeyHex string `json:"public_key_hex"`
	}
	json.Unmarshal(loadJSON("signed-tree-heads.json"), &vectors)

	for _, v := range vectors {
		v := v
		t.Run(v.Description, func(t *testing.T) {
			ok := verifier.VerifyTreeHead(v.TreeSize, b32(v.RootHashHex), v.TimestampNs, h(v.SignatureHex), h(v.PublicKeyHex))
			if !ok {
				t.Fatal("verify_tree_head returned false")
			}
		})
		t.Run(v.Description+" tampered tree_size fails", func(t *testing.T) {
			ok := verifier.VerifyTreeHead(v.TreeSize+1, b32(v.RootHashHex), v.TimestampNs, h(v.SignatureHex), h(v.PublicKeyHex))
			if ok {
				t.Fatal("expected false for tampered tree_size")
			}
		})
		t.Run(v.Description+" tampered root fails", func(t *testing.T) {
			root := b32(v.RootHashHex)
			root[0] ^= 0xFF
			ok := verifier.VerifyTreeHead(v.TreeSize, root, v.TimestampNs, h(v.SignatureHex), h(v.PublicKeyHex))
			if ok {
				t.Fatal("expected false for tampered root")
			}
		})
	}
}

// ── Sequential STH verification ───────────────────────────────────────────────

func TestVerifyTreeHeadSequential(t *testing.T) {
	var vectors []struct {
		Description  string `json:"description"`
		TreeSize     uint64 `json:"tree_size"`
		RootHashHex  string `json:"root_hash_hex"`
		TimestampNs  int64  `json:"timestamp_ns"`
		SignatureHex string `json:"signature_hex"`
		PublicKeyHex string `json:"public_key_hex"`
	}
	json.Unmarshal(loadJSON("signed-tree-heads.json"), &vectors)

	// Consecutive pairs must pass (timestamps are strictly increasing).
	for i := 0; i+1 < len(vectors); i++ {
		prev, curr := vectors[i], vectors[i+1]
		name := fmt.Sprintf("%s after %s", curr.Description, prev.Description)
		t.Run(name, func(t *testing.T) {
			ok := verifier.VerifyTreeHeadSequential(
				curr.TreeSize, b32(curr.RootHashHex), curr.TimestampNs,
				h(curr.SignatureHex), h(curr.PublicKeyHex), prev.TimestampNs,
			)
			if !ok {
				t.Fatal("expected true for sequential pair")
			}
		})
	}

	// Equal timestamp must fail.
	v := vectors[0]
	t.Run("equal timestamp fails", func(t *testing.T) {
		ok := verifier.VerifyTreeHeadSequential(
			v.TreeSize, b32(v.RootHashHex), v.TimestampNs,
			h(v.SignatureHex), h(v.PublicKeyHex), v.TimestampNs,
		)
		if ok {
			t.Fatal("expected false for equal timestamp")
		}
	})
}

// ── Negative cases ────────────────────────────────────────────────────────────

func TestNegativeCases(t *testing.T) {
	var cases map[string]struct {
		LeafHashHex         string   `json:"leaf_hash_hex"`
		LeafIndex           uint64   `json:"leaf_index"`
		TreeSize            uint64   `json:"tree_size"`
		ProofHex            []string `json:"proof_hex"`
		RootHex             string   `json:"root_hex"`
		OldSize             uint64   `json:"old_size"`
		NewSize             uint64   `json:"new_size"`
		OldRootHex          string   `json:"old_root_hex"`
		NewRootHex          string   `json:"new_root_hex"`
		RootHashHex         string   `json:"root_hash_hex"`
		TimestampNs         int64    `json:"timestamp_ns"`
		SignatureHex        string   `json:"signature_hex"`
		PublicKeyHex        string   `json:"public_key_hex"`
		PreviousTimestampNs int64    `json:"previous_timestamp_ns"`
		ExpectedResult      bool     `json:"expected_result"`
	}
	json.Unmarshal(loadJSON("negative-cases.json"), &cases)

	for name, v := range cases {
		name, v := name, v
		if len(name) > 10 && name[:10] == "inclusion_" {
			t.Run(name, func(t *testing.T) {
				got := verifier.VerifyInclusion(b32(v.LeafHashHex), v.LeafIndex, v.TreeSize, proofs(v.ProofHex), b32(v.RootHex))
				if got != v.ExpectedResult {
					t.Fatalf("got %v, want %v", got, v.ExpectedResult)
				}
			})
		} else if len(name) > 12 && name[:12] == "consistency_" {
			t.Run(name, func(t *testing.T) {
				got := verifier.VerifyConsistency(v.OldSize, v.NewSize, proofs(v.ProofHex), b32(v.OldRootHex), b32(v.NewRootHex))
				if got != v.ExpectedResult {
					t.Fatalf("got %v, want %v", got, v.ExpectedResult)
				}
			})
		} else if len(name) > 21 && name[:21] == "tree_head_sequential_" {
			t.Run(name, func(t *testing.T) {
				got := verifier.VerifyTreeHeadSequential(
					v.TreeSize, b32(v.RootHashHex), v.TimestampNs,
					h(v.SignatureHex), h(v.PublicKeyHex), v.PreviousTimestampNs,
				)
				if got != v.ExpectedResult {
					t.Fatalf("got %v, want %v", got, v.ExpectedResult)
				}
			})
		}
	}
}

File: sdks/typescript/tests/verifier.test.ts
import { readFileSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import {
  leafHash,
  nodeHash,
  verifyConsistency,
  verifyInclusion,
  verifyTreeHead,
  verifyTreeHeadSequential,
} from '../src/verifier.js';
import type { SignedTreeHead } from '../src/types.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const VECTORS = resolve(__dirname, '../../../test-vectors');

function loadJson(name: string): unknown {
  return JSON.parse(readFileSync(resolve(VECTORS, name), 'utf-8'));
}

function hex(s: string): Uint8Array {
  return s.length === 0 ? new Uint8Array(0) : new Uint8Array(Buffer.from(s, 'hex'));
}

// ── Leaf hashes ───────────────────────────────────────────────────────────────

describe('leafHash', () => {
  const vectors = loadJson('leaf-hashes.json') as Array<{
    description: string;
    input_hex: string;
    hash_hex: string;
  }>;

  for (const v of vectors) {
    it(v.description, () => {
      const got = leafHash(hex(v.input_hex));
      expect(Buffer.from(got).toString('hex')).toBe(v.hash_hex);
    });
  }
});

// ── Node hashes ───────────────────────────────────────────────────────────────

describe('nodeHash', () => {
  const vectors = loadJson('node-hashes.json') as Array<{
    description: string;
    left_hex: string;
    right_hex: string;
    hash_hex?: string;
    swapped_hash_hex?: string;
  }>;

  for (const v of vectors) {
    if (v.hash_hex) {
      it(v.description, () => {
        const got = nodeHash(hex(v.left_hex), hex(v.right_hex));
        expect(Buffer.from(got).toString('hex')).toBe(v.hash_hex);
      });
    }
    if (v.swapped_hash_hex) {
      it(`${v.description} — non-commutative`, () => {
        const ab = nodeHash(hex(v.left_hex), hex(v.right_hex));
        const ba = nodeHash(hex(v.right_hex), hex(v.left_hex));
        expect(Buffer.from(ab).toString('hex')).not.toBe(Buffer.from(ba).toString('hex'));
        expect(Buffer.from(ba).toString('hex')).toBe(v.swapped_hash_hex);
      });
    }
  }
});

// ── Inclusion proofs ──────────────────────────────────────────────────────────

describe('verifyInclusion', () => {
  const vectors = loadJson('inclusion-proofs.json') as Array<{
    tree_size: number;
    leaf_index: number;
    leaf_hash_hex: string;
    proof_hex: string[];
    root_hex: string;
  }>;

  for (const v of vectors) {
    it(`size=${v.tree_size} idx=${v.leaf_index}`, () => {
      expect(
        verifyInclusion(
          hex(v.leaf_hash_hex),
          BigInt(v.leaf_index),
          BigInt(v.tree_size),
          v.proof_hex.map(hex),
          hex(v.root_hex),
        ),
      ).toBe(true);
    });
  }
});

// ── Consistency proofs ────────────────────────────────────────────────────────

describe('verifyConsistency', () => {
  const vectors = loadJson('consistency-proofs.json') as Array<{
    old_size: number;
    new_size: number;
    old_root_hex: string;
    new_root_hex: string;
    proof_hex: string[];
  }>;

  for (const v of vectors) {
    it(`old=${v.old_size} new=${v.new_size}`, () => {
      expect(
        verifyConsistency(
          BigInt(v.old_size),
          BigInt(v.new_size),
          v.proof_hex.map(hex),
          hex(v.old_root_hex),
          hex(v.new_root_hex),
        ),
      ).toBe(true);
    });
  }
});

// ── Signed Tree Heads ─────────────────────────────────────────────────────────

describe('verifyTreeHead', () => {
  const vectors = loadJson('signed-tree-heads.json') as Array<{
    description: string;
    tree_size: number;
    root_hash_hex: string;
    timestamp_ns: number;
    signature_hex: string;
    public_key_hex: string;
  }>;

  for (const v of vectors) {
    it(v.description, () => {
      const sth: SignedTreeHead = {
        treeSize: BigInt(v.tree_size),
        rootHash: hex(v.root_hash_hex),
        timestampNs: BigInt(v.timestamp_ns),
        signature: hex(v.signature_hex),
        publicKey: hex(v.public_key_hex),
        keyVersion: 1,
      };
      expect(verifyTreeHead(sth)).toBe(true);
    });

    it(`${v.description} — tampered tree_size fails`, () => {
      const sth: SignedTreeHead = {
        treeSize: BigInt(v.tree_size) + 1n,
        rootHash: hex(v.root_hash_hex),
        timestampNs: BigInt(v.timestamp_ns),
        signature: hex(v.signature_hex),
        publicKey: hex(v.public_key_hex),
        keyVersion: 1,
      };
      expect(verifyTreeHead(sth)).toBe(false);
    });

    it(`${v.description} — tampered root_hash fails`, () => {
      const root = Buffer.from(hex(v.root_hash_hex));
      root[0] ^= 0xff;
      const sth: SignedTreeHead = {
        treeSize: BigInt(v.tree_size),
        rootHash: root,
        timestampNs: BigInt(v.timestamp_ns),
        signature: hex(v.signature_hex),
        publicKey: hex(v.public_key_hex),
        keyVersion: 1,
      };
      expect(verifyTreeHead(sth)).toBe(false);
    });
  }
});

// ── Sequential STH verification ───────────────────────────────────────────────

describe('verifyTreeHeadSequential', () => {
  const vectors = loadJson('signed-tree-heads.json') as Array<{
    description: string;
    tree_size: number;
    root_hash_hex: string;
    timestamp_ns: number;
    signature_hex: string;
    public_key_hex: string;
  }>;

  for (let i = 0; i + 1 < vectors.length; i++) {
    const prev = vectors[i];
    const curr = vectors[i + 1];
    it(`${curr.description} after ${prev.description}`, () => {
      const sth: SignedTreeHead = {
        treeSize: BigInt(curr.tree_size),
        rootHash: hex(curr.root_hash_hex),
        timestampNs: BigInt(curr.timestamp_ns),
        signature: hex(curr.signature_hex),
        publicKey: hex(curr.public_key_hex),
        keyVersion: 1,
      };
      expect(verifyTreeHeadSequential(sth, BigInt(prev.timestamp_ns))).toBe(true);
    });
  }

  it('equal timestamp fails', () => {
    const v = vectors[0];
    const sth: SignedTreeHead = {
      treeSize: BigInt(v.tree_size),
      rootHash: hex(v.root_hash_hex),
      timestampNs: BigInt(v.timestamp_ns),
      signature: hex(v.signature_hex),
      publicKey: hex(v.public_key_hex),
      keyVersion: 1,
    };
    expect(verifyTreeHeadSequential(sth, BigInt(v.timestamp_ns))).toBe(false);
  });
});

// ── Negative cases ────────────────────────────────────────────────────────────

describe('negative cases', () => {
  const all = loadJson('negative-cases.json') as Record<
    string,
    {
      leaf_hash_hex?: string;
      leaf_index?: number;
      tree_size?: number;
      proof_hex?: string[];
      root_hex?: string;
      old_size?: number;
      new_size?: number;
      old_root_hex?: string;
      new_root_hex?: string;
      root_hash_hex?: string;
      timestamp_ns?: number;
      signature_hex?: string;
      public_key_hex?: string;
      previous_timestamp_ns?: number;
      expected_result: boolean;
    }
  >;

  for (const [name, v] of Object.entries(all)) {
    if (name.startsWith('inclusion_')) {
      it(name, () => {
        expect(
          verifyInclusion(
            hex(v.leaf_hash_hex!),
            BigInt(v.leaf_index!),
            BigInt(v.tree_size!),
            v.proof_hex!.map(hex),
            hex(v.root_hex!),
          ),
        ).toBe(v.expected_result);
      });
    } else if (name.startsWith('consistency_')) {
      it(name, () => {
        expect(
          verifyConsistency(
            BigInt(v.old_size!),
            BigInt(v.new_size!),
            v.proof_hex!.map(hex),
            hex(v.old_root_hex!),
            hex(v.new_root_hex!),
          ),
        ).toBe(v.expected_result);
      });
    } else if (name.startsWith('tree_head_sequential_')) {
      it(name, () => {
        const sth: SignedTreeHead = {
          treeSize: BigInt(v.tree_size!),
          rootHash: hex(v.root_hash_hex!),
          timestampNs: BigInt(v.timestamp_ns!),
          signature: hex(v.signature_hex!),
          publicKey: hex(v.public_key_hex!),
          keyVersion: 1,
        };
        expect(verifyTreeHeadSequential(sth, BigInt(v.previous_timestamp_ns!))).toBe(
          v.expected_result,
        );
      });
    }
  }
});
